### ISSUE TITLE
Testing Text-Heavy Exercises

### ISSUE THREAD

**Issue by tstirrat15** [exercism/xpython#168]
I'm working on the Twelve Days problem right now, and it's reminding me of something that's been annoying me as I've worked through the exercises.

The default AssertionError that prints to the terminal contains a diff of the two bodies of text, but the diff is hard to parse when you're dealing with large bodies of text, and what's extra/missing from your program's output isn't immediately apparent.

Is there another way to present test output for these kinds of exercises? Is there some extension of UnitTest that gives more useful feedback?

**Comment by kytrinyx** [comment]
That's a really good question, and it's a problem in all of the tracks. I don't have answers, I just wanted to say that I'm interested in hearing if other people have useful input on this.

**Comment by tstirrat15** [comment]
One thing that sounds like it could be fairly robust across languages is to have the test write the output of each function to a local file in some standard format, and then call an external diff utility.

It'd add another hoop for the user to jump through, in that they'd have to have the diff utility available.

**Comment by kytrinyx** [comment]
Mac and linux boxes would always have the basic `diff` available, does windows have something like that?

**Comment by tstirrat15** [comment]
I doubt it, but I don't know. One thought: is `git` a part of the backbone of `exercism`? Because then there should be a default `difftool` that you can call using `git`.

**Comment by kytrinyx** [comment]
No, there's no assumption that people use git.

**Comment by outkaj** [comment]
It looks like there is a built-in equivalent of diff for Windows, called FC (FC file1 file2 would be the equivalent of diff file1 file2). I'm running Linux, so I can't verify this, but could someone running Windows confirm?

**Comment by kytrinyx** [comment]
@jwood803 @LegalizeAdulthood - would you be able to verify this?

**Comment by tstirrat15** [comment]
I just checked it, and it exists.

It's a very simple command-line utility, and doesn't seem like it would be much more useful than the output of the tests.

[Here's](https://gist.github.com/tstirrat15/7de75c79708bd965802f) what it returned when I was playing with it.

**Comment by jwood803** [comment]
A quick search didn't turn up anything useful in terms of the .NET crowd. I'll keep at it to see if there's a solution for this in terms of Windows other than downloading an actual merge tool lie WinMerge.

**Comment by kytrinyx** [comment]
Now that I think of it... @isodore do you happen to know what the inherent diffing capabilities of windows are? I imagine that you've been down this route with approvals tests.

**Comment by isidore** [comment]
Yes this is exactly approvaltests. There are a host of options for windows. The best would be to look at the diffreporter in either the Java or c# version of approval tests. Will send link when not on my phone...

Llewellyn Falco

**Comment by isidore** [comment]
https://github.com/approvals/ApprovalTests.Java/tree/master/java/org/approvaltests/reporters/windows

but this is even more complete

https://github.com/approvals/ApprovalTests.Net/tree/master/ApprovalTests/Reporters

**Comment by kytrinyx** [comment]
Sweet, thanks!

**Comment by LegalizeAdulthood** [comment]
**Note:** The approvals test framework isn't implemented for all the language tracks on exercism.

**Comment by kytrinyx** [comment]
Yeah, not even close, but some of the ideas there could help us figure out what to do with the text-heavy exercises.

**Comment by LegalizeAdulthood** [comment]
Yeah, I sometimes wish that unit test frameworks in general would act more like diff tools when comparing text.  It's particularly annoying when the difference is due to misplaced whitespace and you're having a real hard time seeing it in output that says "a really big long interesting string with lots of words in it" !+ "a really biglong interesting string with lots of words in it"

**Comment by kytrinyx** [comment]
I have lost SO much time due to differences in trailing whitespace.

**Comment by behrtam** [comment]
The default python unittest diff is not so bad as it underlines the missing comma and whitespace in my example.

```
======================================================================
FAIL: test_multi_line_equal (__main__.TestStringMethods)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_line.py", line 22, in test_multi_line_equal self.assertMultiLineEqual(TEXT_A, TEXT_B)
AssertionError: 'Lorem ipsum dolor sit amet, consetetur sadipscing elitr,\nsed diam nonumy eirmo [truncated]...
- Lorem ipsum dolor sit amet, consetetur sadipscing elitr,
?                                                        -
+ Lorem ipsum dolor sit amet, consetetur sadipscing elitr
  sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat,
  sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum.
- Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
?                           -
+ Stet clita kasd gubergren,no sea takimata sanctus est Lorem ipsum dolor sit amet.

----------------------------------------------------------------------
Ran 1 tests in 0.001s
```

**Comment by kytrinyx** [comment]
Oh, wow. The underlines help a lot.
